### PR TITLE
ENG-1932 - Update Unity SDK

### PR DESCRIPTION
**Pull request by DavidYangSphere1**
### **PR Description**

Jira Ticket: [ENG-1932](https://sphereone.atlassian.net/browse/ENG-1932)

This PR has changes to update the unity codebase to have parity with the SphereOne WebSDK.

**NOTE**: Please be aware that some changes in the `SphereOneManager.cs` file and maybe other places have code modifications that should not be present in the final PR. I've made some updates, so I can hit a local testing environment, instead of production. Once done, will revert those changes.

**NOTE**: Also, I've rebuilt the unity sdk in `Unity v2022.3.4f1`. Previously, we built the Unity SDK in `Unity v2022.3.0f1`. But I had to build with a higher version because of Xcode 15 for iOS. You can find more info in [this link here](https://developer.apple.com/forums/thread/737871) and [here](https://issuetracker.unity3d.com/issues/building-projects-with-il2cpp-scripting-backend-for-apple-platforms-fails-with-xcode-15-dot-0b6-or-newer?page=2#comments). I don't know which forum thread or reddit post I read previously that said that the latest version of Unity had resolved the issue. But anyways, this will be something that I would need to include in the client-facing documentation anyways.

### **ChangeLogs**

- added more data types in `ApiTypes.cs` with similar names as WebSDK. There are some differences due to the limited or additional capabilities in C# Unity as compared to javascript/typescript
- refactored the `WebRequestHandler.cs` (overloading functions when necessary per case) and have proper error handling mechanisms. Previously, if there was an error, simply returned `null`, to minimize the need of `try-catch`. Now, it's actually returning proper errors
- updated `SphereOneManager.cs` to use the new methods within `WebRequestHandler`, have `try-catch` blocks, handle custom errors for `payCharge` and `getRouteEstimation`, added `nftTransfers`, and resolve issues due to these big changes. Also, in the `catch` block, I'm just returning `null` a lot for "backward compatibility" but will probably change them and just throw the error or a message
- fix compilation issues inside of Unity
- still more to come, once `payCharge` and `nftTransfers` are tested on web, mobile, and desktops

### **Additional Notes**

```
This is referring to the Demo Project in the Unity SDK itself and
NOT, I repeat "NOT", the Car Shooter Unity Demo Game.

iOS demo on unity-sdk just broke on me as I was trying to get a demo video for iOS.
It is not the code of the implementation. That, I can assure you, since it was working
before I did any major updates on my macbook over Christmas.
It is unity's compatibility with Xcode when ejecting an Xcode iOS project setup.

The unity version that I was using, was working, as mentioned in the
wordy PR description.
But I guess, it's no longer compatible and I need to fix that now. But that doesn't
involve making any code changes.
I think updating to some latest version of Unity should resolve that.
```

### **Showcase**

```
I am so sorry. I talk so much. 🙇 

The video clips below are actually from 1 giant video clip, roughly 1 hr+ long. 🙇 
I did cut out some contents, mostly me complaining 😞 , but I wanted to give out
as much info as I can.
```
<table>
   <tr>
       <td colspan="1" align="center"><b>Explaining my code changes</b></td> 
   </tr>
   <tr>
       <td colspan="1" align="center"><b>Intro and Me Complaining </b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/457d7f85-29e3-415f-a753-81039bc47f07" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>DemoScene</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/1314db4b-dfb7-42f9-bc5d-a32c2304efc6" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>JS Files</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/5c49a8df-9024-4ff5-b9b7-c68a14663edd" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>Api Call File</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/1f8f92ca-c214-497e-86f9-d1f74518e4ee" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>Data Type File</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/e23d46fd-a88f-48dc-9c0f-ced74e78742e" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>Actual SDK Updates Part 1</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/eb548d69-9c97-4eb4-b937-2b57b49a7a17" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>Actual SDK Updates Part 2</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/9ab70428-9589-4d43-af3c-996708a1b606" controls="controls" muted="muted"></video></td>
   </tr>
   <tr>
       <td colspan="1" align="center"><b>PinCode/Login Reuse</b></td> 
   </tr>
   <tr>
       <td align="center"><video src="https://github.com/SphereGlobal/unity-sdk/assets/116052081/ebeebc98-6a0d-45b4-b92e-fc127777b9c8" controls="controls" muted="muted"></video></td>
   </tr>
</table>

[ENG-1932]: https://sphereone.atlassian.net/browse/ENG-1932?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 11 14 48 AM

```csharp
public class WebRequestResponse
{
    public string Data { get; set; }
    public string Error { get; set; }
    public bool IsSuccess { get; set; }

    public WebRequestResponse(string data, string error, bool isSuccess)
    {
        Data = data;
        Error = error;
        IsSuccess = isSuccess;
    }
}
```

For this suggestion, I am already returning the error from the api response. The reason it is a `string` because the `UnityWebRequest` returns a `string` for the data response and error response. It's stringified.

Thus, the actual error parsing is handled in the `SphereOneManager.cs` file.
If it is suggesting that I parse the error here in the `WebRequestHandler`, that can be done, but this class is supposed to handle api requests, and the `SphereOneManager.cs` is the middle-layer that handles those errors.

So, I don't think this is required.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 11 14 52 AM

This should already be private.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 11 15 01 AM

This suggestion is proper. While I would like to make changes here, this `TestSphereOneManager` is a demo. I'm already console-logging the error.

And this function `TransferNft` only does that, demoing the functionality to transfer an NFT. It doesn't do anything else. If there's no nfts, I throw an error in order to console-log it. I don't know what other conditions should I do?

If I did `return`, user wouldn't know what happened.

I could show Alert Dialogs, but they're only built-in functionality for the Unity Editor, while WebGL, Mobile, and Desktop, I would need to implement custom logic. So, I think that's doing too much.

For now, I'll leave it as it is.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 4 10 41 PM

For this code suggestion, I asked Cursor's gpt bot, and this is what it says:

The line `await Task.Yield().ConfigureAwait(false);` is used in asynchronous programming in C# to achieve a couple of things:
1. `await Task.Yield();` is a way to asynchronously yield control back to the caller of the method. This will schedule the remainder of the method to continue after the current synchronous block of code has completed and the call stack has unwound. It's often used to ensure that the UI remains responsive by not running a long-running operation on the UI thread.

2. `ConfigureAwait(false)` configures the task to continue on a thread pool thread rather than capturing and resuming on the original context (which, in a UI application, would be the UI thread). This is important for avoiding deadlocks in GUI applications when the continuation of an async method tries to run on a thread that is waiting for the async method to complete.

Here's a breakdown of when and why you might use this:

- `await Task.Yield();`: This is useful when you have an async method that starts with a potentially long-running synchronous block of code. By yielding, you ensure that the initial part of the method completes quickly, and the rest of the work is scheduled to continue later, allowing the UI thread to remain responsive.

- `ConfigureAwait(false)`: This is typically used in library code where you don't need to resume execution on the original context. It can help prevent deadlocks and improve performance by not forcing the continuation to marshal back to the original context (like the UI thread).

Combining them as `await Task.Yield().ConfigureAwait(false);` is somewhat unusual. It's not incorrect per se, but it's important to understand what it does and why you might need it. In most cases, you would use `ConfigureAwait(false)` on its own with other `awaitable` tasks to avoid capturing the synchronization context.

....So, in summary, if I use `ConfigureAwait(false)`, I can improve the asynchronous task, but it's primarily used with other asynchronous tasks and doesn't need to go back to the mainThread.

Whereas, `await Task.Yield()` allows the asynchronous call to route back to the mainThread for UI to resume, so a bit of headway since it needs a reference to the mainThread.

In other words, I think I will keep using `Task.Yield()` here, because it fits more for our case, since we're going to be used on frontend for the most part.

If I was writing a backend service in .NET, I may have chosen the `await Task.Yield().ConfigureAwait(false)`. And like, if I needed to do any intense async tasks that doesn't require the result to be sent back to the mainThread, then I guess that would also work too.

My only concern is, because the demo `TestSphereOneManager` is so simple, without much UI activities, I may not be seeing any performance or unforeseen issues. So, I'll probably keep this in mind for now, and enable it when performance issues are brought up.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 4 26 50 PM

While this code suggestion is good, I don't think I'll make this change. This is an internal function, which I don't think anyone else will use it like that.
Unless they use the `UnityWebRequest` directly in a new function. They shouldn't need to directly input a `PUT` or `POST` into the function.

**Comment by DavidYangSphere1** [comment]
/improve --pr_code_suggestions. num_code_suggestions=8 --pr_code_suggestions. extra_instructions="focus on SphereOneManager.cs file, offer suggestions for functions for CreateCharge, Pay, GetRouteEstimation, and TransferNft"

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 5 46 36 PM

While it is good practice, this is a demo. And I'm directly opening a website here, so I think it's only fair to simply do `return` instead of a `throw` here.

As for the other functions in this file, I think those, except for `GetRouteEstimation`, should have errors thrown.
But as I mentioned previously, this is a demo. I'm already logging, so I think that's sufficient enough.

If it was an actual implementation file, then more considerations would be put.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 6 01 13 PM

Yes, this would make sense. But currently, in Unity, this function is directly attached to a UI component. Upon press, a clickEvent triggers and calls this function directly.

Currently, there's no UI popup for users to input a receiver address, which could be passed into this function.
So, will not implement this.

**Comment by DavidYangSphere1** [comment]
Screenshot 2024-01-25 at 7 02 49 PM

Well, I did try it out since I was talking with Cursor's gpt bot. It says that, `Task.Yield()` and `Task.Delay()` should do the same thing and not block the mainThread.

But when I did add it in, the api calls are being made and UI isn't blocked. But the async response isn't being returned back to the mainThread, probably still in the background thread.

So, I won't implement this one.

**Comment by DavidYangSphere1** [comment]
/improve --pr_code_suggestions.extra_instructions="SphereOneManager.cs file"

**Review by ryanSphereOne — APPROVED** [review]
Please be good
